include/zephyr/async.h

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Public APIs for Async utilities
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous functions completion.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ASYNC_H_
+#define ZEPHYR_INCLUDE_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous callbacks
+ * and implemented for some commonly used IPC mechanisms.
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
+
+/**
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
```

Comment:
Why not having 1 structure for all, and with a name like "async_notifier"?

Sure, k_poll_signal is not going to use the result attribute but it's only 4 bytes but it's not like there will be hundreds of such objects.

And then would be nice to have macros to set up each type of notifier (so the user does not have to go through the burden of manipulating sem/mutex/k_work/etc... by himself).
Perhaps also a macro or iniline as a helper to give the right generic callback for a selected type of notifier. 

Idea is that it could be made easy in user code to switch to another notifier.

Reply:
I'll consider what this means